sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -36,6 +37,33 @@ def _pack(coefs_, intercepts_):
     return np.hstack([l.ravel() for l in coefs_ + intercepts_])
 
 
+def _safe_sparse_elementwise_multiply(a, b):
+    """element-wise multiplication that deals with sparse matrix
+    """
+    if issparse(a):
+        return a.multiply(b)
+    elif issparse(b):
+        return b.multiply(a)
+    else:
+        return a * b
+
+
+def _efficient_dropout_mask_sample(X, retain_prob, random_state):
+    """Efficient way of sampling dropout mask, given X could be sparse matrix
+
+    If X is sparse matrix, we should sample only as many times as number of
+    non-zero values in X.
+    """
+    if issparse(X):
+        sampled_values = random_state.binomial(
+            1, retain_prob, X.nnz) / retain_prob
+        X_coo = X.tocoo()
+        return coo_matrix((sampled_values, (X_coo.row, X_coo.col)),
+                          shape=X.shape)
+    else:
+        return random_state.binomial(1, retain_prob, X.shape) / retain_prob
```

Comment:
@sergulaydore right, one can either do `output * p_retain` at prediction time or do `output / p_retain` at training time. The latter seems more common.